content/BCPatterns/generic-method-pattern/index.md

```diff
@@ -0,0 +1,327 @@
+
+```AL
+codeunit 53100 "WLD BlockCustomer Meth"
+{
+    internal procedure BlockCustomer(var Cust: Record Customer; HideDialog: Boolean);
+    var
+        Handled: Boolean;
+    begin
+        if not ConfirmBlockCustomer(HideDialog) then exit;
+        ...
+        AcknowledgeBlockCustomer(HideDialog)
+    end;
+...
+
+    local procedure ConfirmBlockCustomer(HideDialog: Boolean): Boolean
+    var
+        ConfirmManagement: Codeunit "Confirm Management";
+        ConfirmQst: Label 'Are you sure?';
+        DefaultAnswer: Boolean;
+    begin
+        DefaultAnswer := true;
+
+        if HideDialog then exit(DefaultAnswer);
+        exit(ConfirmManagement.GetResponseOrDefault(ConfirmQst, DefaultAnswer));
+    end;
+
+    local procedure AcknowledgeBlockCustomer(HideDialog: Boolean)
+    var
+        AcknowledgeMsg: Label 'You successfully executed "BlockCustomer"';
+    begin
+        if not GuiAllowed or HideDialog then exit;
+        Message(AcknowledgeMsg);
+    end;
+...
+}
+```
+
+*The Event layer*
+This layer is going to add flexibility to any app that has a dependency on this app.  By default, the pattern always foresees an `OnBefore` and an `OnAfter` event.  
+
+This is the relevant code for the event layer:
+
+```AL
+codeunit 53100 "WLD BlockCustomer Meth"
+{
+    internal procedure BlockCustomer(var Cust: Record Customer; HideDialog: Boolean);
+    var
+        Handled: Boolean;
+    begin
+        ...
+        OnBeforeBlockCustomer(Cust, Handled);
+        ...
+        OnAfterBlockCustomer(Cust);
+        ...
+    end;
+...    
+    [IntegrationEvent(false, false)]
+    local procedure OnBeforeBlockCustomer(var Cust: Record Customer; var Handled: Boolean);
+    begin
+    end;
+
+    [IntegrationEvent(false, false)]
+    local procedure OnAfterBlockCustomer(var Cust: Record Customer);
+    begin
+    end;
+}
+```
+
+*The method layer*
+The last layer is obviously where the business logic will be written.
+
+The relevant part is:
+
+```AL
+codeunit 53100 "WLD BlockCustomer Meth"
+{
+    internal procedure BlockCustomer(var Cust: Record Customer; HideDialog: Boolean);
+    var
+        Handled: Boolean;
+    begin
+        ...
+        DoBlockCustomer(Cust, Handled);
+        ...
+    end;
+
+    local procedure DoBlockCustomer(var Cust: Record Customer; Handled: Boolean);
+    begin
+        if Handled then
+            exit;
+
+        Cust.Blocked := Cust.Blocked::All;
+        Cust.Modify(true);
+    end;
+...
+}
+```
+Usually indicated with a "do"-function, the business logic takes place in that procedure.  Obviously, when you have a decent amount of code, it's recommended that you make it readable by applying all the Best Practices in terms of readability in the codeunit.  Though, a few pointers here:
+- keep the cyclomatic complexity low
```

Comment:
Maybe add a link to https://en.wikipedia.org/wiki/Cyclomatic_complexity or add a (very) short abstract how it's calculated